riscv/src/code_gen.rs

```diff
@@ -319,43 +376,179 @@ fn preamble<T: FieldElement>(runtime: &Runtime, with_bootloader: bool) -> String
         + &bootloader_preamble_if_included
         + &memory(with_bootloader)
         + r#"
-    // ============== Constraint on x0 =======================
-
-    x0 = 0;
-
-    // ============== iszero check for X =======================
-    let XIsZero = std::utils::is_zero(X);
+    // =============== Register memory =======================
+"# + format!("std::machines::memory::Memory_{} regs;", degree + 2)
+        .as_str()
+        + r#"
+    // Get the value in register Y.
+    instr get_reg Y -> X link ~> X = regs.mload(Y, STEP);
+
+    // Set the value in register X to the value in register Y.
+    instr set_reg X, Y -> link ~> regs.mstore(X, STEP, Y);
+
+    // We still need val1 and val2 for prover inputs.
+    reg val1;
+    reg val2;
```

Comment:
Technically we could, but it's gonna get hacky. Currently the extra registers get added automatically if poseidon or arith are used, which is not the "base" machine. Although this is more costly, I think it's cleaner because of the clear separation. We should just fix the syscall stuff separately. I'll change the names for now.